In [408]:
import numpy as np




## 1

In [418]:
c=np.array([-2,-5,0,0,0]); # coefficents of objective
A=np.array([[1,0,1,0,0],[0,1,0,1,0],[1,1,0,0,1]]); # constraint matrix
x0=np.array([2,3,2,3,3]);
X=np.diag(x0)
l=np.array([1,1,1,1,1]) # e vector
mui=2 # 
rho=0.9999 # decay factor
a=0.3 # alpha
i=0 # iterate number
max_steps=10000
x=x0
lamb=np.matmul(np.linalg.inv(np.matmul(np.matmul(A,np.matmul(X,X)),A.T)),np.matmul(np.matmul(A,np.matmul(X,X)),c))-mui*np.matmul(np.linalg.inv(np.matmul(np.matmul(A,np.matmul(X,X)),A.T)),np.matmul(A,x));
B=lambda x : np.dot(c,x)-mui*np.sum(np.log(x)) # log Barrier objective function
B1=lambda x : np.dot(c,x) # Original objective function


In [390]:
while np.linalg.norm((np.matmul(X,c-np.matmul(A.T,lamb)))-mui*l)>=1e-4:
    
    mui=rho*mui;
    X=np.diag(x)
    lamb=np.matmul(np.linalg.inv(np.matmul(np.matmul(A,np.matmul(X,X)),A.T)),np.matmul(np.matmul(A,np.matmul(X,X)),c))-mui*np.matmul(np.linalg.inv(np.matmul(np.matmul(A,np.matmul(X,X)),A.T)),np.matmul(A,x));
    # lambda vector
    p=x+(1/mui)*np.matmul(np.matmul(X,X),np.matmul(A.T,lamb)-c); # Newton Direction
    a=rho*a # naive alpha decayed by factor rho
    x=x+a*p # update
    i=i+1
    if i%36013==0:
        
        n=np.linalg.norm((np.matmul(X,c-np.matmul(A.T,lamb)))-mui*l)
        print('After {} iterations'.format(i))
        print('The projected coordinates are ({},{}).'.format(x[0],x[1]))
   
        print('The value of the log barrier function is {}'.format(B(x)))
    
        print(n) # value of error; for comparing to stopping criterion
    
    
    
    
    
    
   
    
    
    
    
    
    
    

After 36013 iterations
The projected coordinates are (1.9907452901284306,5.981648251850886).
The value of the log barrier function is -33.648942055778235
0.0009302310262879217
After 72026 iterations
The projected coordinates are (1.9996650355393155,5.999330294862366).
The value of the log barrier function is -33.97955484530018
0.0007363942893243075
After 108039 iterations
The projected coordinates are (1.9999541648366703,5.999908333935491).
The value of the log barrier function is -33.99884027081019
0.00033148067022383443
After 144052 iterations
The projected coordinates are (1.9999809532487236,5.999961907261851).
The value of the log barrier function is -33.999752861109435
0.00016005003142948517
After 180065 iterations
The projected coordinates are (1.999988209864074,5.999976420054559).
The value of the log barrier function is -33.999857984027386
9.999999389249191e-05


## Comments

To satisfy the positivity of $\vec x_{k}=\vec x_{k-1}+\alpha_{k}\vec p_{k}$, a naive strategy was chosen for getting a sequence $\alpha_{k}$ so that all $\vec x_{k}$ are positive for all iterates and converges to the minimum. The parameters $a,rho,mui$ had to be tuned as a result. Looking at the results of the iterations, the minimum appears to be  at $(2,6)$. These are the projected coordiantes obtained by only taking the first two values corresponding to the $x,y$ plane.The value of the log Barrier function converges slowly to $-34$; the minimum value is $-34$. After $180,065$ iterations, the stopping criterion is satsified with chosen set of parameters($a=0.3$, $mui=2$, $rho=0.999$). Of course, it is possible to converge faster using other set of parameters(but has to be adjusted to satisfy stopping criterion). I only printed a few of the iterations which includes the final one.

The bottom code shows a smaller value of rho that converges faster but the stopping criterion is ignored but you still get the same minimizer in less iterations.

In [336]:
rho=0.9 # smaller rho


In [337]:
while np.linalg.norm((np.matmul(X,c-np.matmul(A.T,lamb)))-mui*l)>=1e-4:
    
    mui=rho*mui;
    X=np.diag(x)
    lamb=np.matmul(np.linalg.inv(np.matmul(np.matmul(A,np.matmul(X,X)),A.T)),np.matmul(np.matmul(A,np.matmul(X,X)),c))-mui*np.matmul(np.linalg.inv(np.matmul(np.matmul(A,np.matmul(X,X)),A.T)),np.matmul(A,x));
    # lambda vector
    p=x+(1/mui)*np.matmul(np.matmul(X,X),np.matmul(A.T,lamb)-c); # Newton Direction
    a=rho*a # naive alpha decayed by factor rho
    x=x+a*p # update
    i=i+1
    if i%100==0: # print every 100 iterations
        
        n=np.linalg.norm((np.matmul(X,c-np.matmul(A.T,lamb)))-mui*l)
        print('After {} iterations'.format(i))
        print('The projected coordinates are ({},{}).'.format(x[0],x[1]))
   
        print('The value of the log barrier function is {}'.format(B(x)))
    
        print(n) # value of error; for comparing to stopping criterion

After 100 iterations
The projected coordinates are (1.9871877884433669,5.974629729094857).
The value of the log barrier function is -33.84732417357981
0.10900005949039372
After 200 iterations
The projected coordinates are (1.994072129976868,5.988202274753383).
The value of the log barrier function is -33.92915562624719
0.05036330462538888
After 300 iterations
The projected coordinates are (1.9961400047433933,5.9923050730501775).
The value of the log barrier function is -33.95380537473744
0.03277842221140264
After 400 iterations
The projected coordinates are (1.997137380439042,5.994288669335955).
The value of the log barrier function is -33.965718107557855
0.024303579337443874
After 500 iterations
The projected coordinates are (1.9977248599570658,5.995458551600463).
The value of the log barrier function is -33.97274247791645
0.019313528284809724
After 600 iterations
The projected coordinates are (1.9981121348746167,5.9962303716471).
The value of the log barrier function is -33.977376127

After 5000 iterations
The projected coordinates are (1.9997773627899627,5.999554811705071).
The value of the log barrier function is -33.99732878410528
0.001889213441200062
After 5100 iterations
The projected coordinates are (1.9997817370671258,5.999563556911455).
The value of the log barrier function is -33.997381258691526
0.0018520935764204187
After 5200 iterations
The projected coordinates are (1.9997859427526767,5.999571965126557).
The value of the log barrier function is -33.997431711138134
0.001816404424102219
After 5300 iterations
The projected coordinates are (1.9997899894097475,5.999580055463148).
The value of the log barrier function is -33.99748025613523
0.0017820648271897623
After 5400 iterations
The projected coordinates are (1.9997938858914253,5.999587845610535).
The value of the log barrier function is -33.99752699983553
0.0017489996531580372
After 5500 iterations
The projected coordinates are (1.9997976404055988,5.99959535197787).
The value of the log barrier function i

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: RuntimeWarning: invalid value encountered in matmul
  """
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: RuntimeWarning: invalid value encountered in matmul
  import sys


## 2

In [457]:
# 2
D=np.eye(5)
lambd=[-5,-3,-4] # chosen so that (x,s)>0
x0=np.array([2,3,2,3,3])
x=x0
b=np.array([4,6,8])
s=-np.matmul(A.T,lambd)+c
S=np.diag(s)# S matrix

#Initial x vector
xl1=np.append(x,lambd)
xl=np.append(xl1,s)

rho=0.5
mui=0.9
i=0 # iteration number
X=np.diag(x) # X matrix

# making initial F vector
e=np.array([1,1,1,1,1])
v=-(np.matmul(A,x)-b)
v1=-(np.matmul(X,s)-mui*e)
F1=np.append(-(np.matmul(A.T,lambd)+s-c),v)
F=np.append(F1,v1)

#Construct block matrix
M=np.block([[np.zeros((5,5)),A.T, D ],[A,np.zeros((3,3)),np.zeros((3,5))],[S,np.zeros((5,3)),X]])

# check that all shapes and vectors are correct

print(s)
print(v.shape)
print(v1.shape)
print(F)
print(F.shape)
print(M.shape)
print(xl.shape)
print(xl)

[7 2 5 3 4]
(3,)
(5,)
[  0.    0.    0.    0.    0.    0.    0.    0.  -13.1  -5.1  -9.1  -8.1
 -11.1]
(13,)
(13, 13)
(13,)
[ 2  3  2  3  3 -5 -3 -4  7  2  5  3  4]


In [458]:
while 5*mui>=1e-4: # stopping criterion
    mui=rho*mui
    a=0.9 # alpha
    D=np.eye(5) # for block matrix M
    X=np.diag(x) # Construct X matrix
    
    s=-np.matmul(A.T,lambd)+c # s values
    
    # x vector containing updated x, lambda, s
    xl1=np.append(x,lambd)
    xl=np.append(xl1,s)
    
    S=np.diag(s) # S matrix
    
    # F vector in lecture
    v=-(np.matmul(A,x)-b)
    v1=-(np.matmul(X,s)-mui*e)
    F1=np.append(-(np.matmul(A.T,lambd)+s-c),v)
    F=np.append(F1,v1)
    
    # Construct block matrix
    M=np.block([[np.zeros((5,5)),A.T, D ],[A,np.zeros((3,3)),np.zeros((3,5))],[S,np.zeros((5,3)),X]])
    
    F2=np.linalg.solve(M,F) #Newton Direction
    a=rho*a # naive alpha decay every iteration
    xl=xl+a*F2 # update
    x=xl[:5] # retrieve x values
    lambd=xl[5:8] # retrieve lambda values
    
    i=i+1
    print('After {} iterations'.format(i))
    print('The projected coordiantes are({},{}) with s>0'.format(x[0],x[1]))
    print('The lambda vector is {}'.format(lambd)) # See lambda every iteration
    print('The s vector is {}'.format(s)) # See that s>0 for every iteration
  

After 1 iterations
The projected coordiantes are(1.9796016483516483,3.736565934065934) with s>0
The lambda vector is [-2.80025412 -2.45406593 -3.22239011]
The s vector is [7 2 5 3 4]
After 2 iterations
The projected coordiantes are(1.9069288888950444,4.5924641334020615) with s>0
The lambda vector is [-1.48952985 -2.32245318 -2.92174562]
The s vector is [4.02264423 0.67645604 2.80025412 2.45406593 3.22239011]
After 3 iterations
The projected coordiantes are(1.8028485975119182,5.2483046937546485) with s>0
The lambda vector is [-0.76935983 -2.39546214 -2.71499715]
The s vector is [2.41127547 0.2441988  1.48952985 2.32245318 2.92174562]
After 4 iterations
The projected coordiantes are(1.7437375797194785,5.637750825284624) with s>0
The lambda vector is [-0.41397005 -2.59224403 -2.46513503]
The s vector is [1.48435699 0.1104593  0.76935983 2.39546214 2.71499715]
After 5 iterations
The projected coordiantes are(1.7732999951770534,5.824181005377598) with s>0
The lambda vector is [-0.23871691 -

## Comments

This algorithm converges to $(1.99,5.99)$ or $(2,6)$(using projected coordinates in $x,y$ plane) which is the location of the same minimizer as seen in the first problem.The same conclusion is reached from the first algorithm. It only took $16$ iterations until the $mui$ became small enough to satisfy the stopping criterion.If we look at the coordinates in the $x,y$ plane at each iteration, we can guess that the path of convergence toward the minimizer is curved,like a horn. The paramteres chosen for this model were :($rho=0.5$, $mui=0.7$,$a=0.9$,).I also chose $\vec x_{0}=(2,3,2,3,3)$ and $\vec \lambda_{0}=(-5,-3,-4)$ had to be chosen to ensure positivity throughout the iterations.

